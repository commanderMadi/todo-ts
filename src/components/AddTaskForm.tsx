import React from 'react';
import { Task, Category } from '../actions';
import uuid from 'uuid';
import { FormContainer, Form } from './styles/forms';
import { Button } from '../components/styles/base';

interface AddTaskFormProps {
  categories: Category[];
  onSubmit(newTask: Task): void;
}

export class AddTaskForm extends React.Component<AddTaskFormProps> {
  constructor(props: AddTaskFormProps) {
    super(props);

    this.onSubmit = this.onSubmit.bind(this);
  }
  state: Task = {
    id: 'default-id',
    title: '',
    categoryTitle: '',
    error: null
  };

  onSubmit(e: any) {
    e.preventDefault();
    let title = e.target.elements[0].value;
    let categoryTitle = e.target.elements[1].value;
    let id = uuid();
    if (title && categoryTitle !== 'Select an option') {
      this.setState(
        () => {
          return {
            title,
            categoryTitle,
            id,
            error: null
          };
        },
        () => this.props.onSubmit(this.state)
      );
      e.target.elements[0].value = '';
      e.target.elements[1].value = 'Select an option';
    } else if (
      !title ||
      !categoryTitle ||
      categoryTitle === 'Select an option'
    ) {
      this.setState(() => {
        return {
          error:
            'Please make sure you you have entered a Task title and selected a category'
        };
      });
    }
  }
  render() {
    return (
      <FormContainer>
        {this.state.error && (
          <p className='col-md-12 text-danger mt-4'>{this.state.error}</p>
        )}
        <h4 className='mt-2'>Add a New Task</h4>
        <Form className='mt-4' onSubmit={this.onSubmit}>
          <input
            className='form-control'
            id='task_title'
            placeholder='Enter task title...'
            type='text'
          />
          <select id='formcontrolselect' className='form-control'>
            <option>Select an option</option>
            {this.props.categories.map((category, i: number) => {
              return <option key={i}>{category.name}</option>;
            })}
          </select>
          <Button className='btn btn-primary' type='submit'>
            Add Task
          </Button>
        </Form>
      </FormContainer>
    );
  }
}
