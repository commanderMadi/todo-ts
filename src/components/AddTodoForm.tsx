import React from 'react';
import { connect } from 'react-redux';
import { Todo, addTodo, Category } from '../actions';

interface AddTodoFormProps {
    categories: Category[];
    addTodo: typeof addTodo;
}

class AddTodoForm extends React.Component<AddTodoFormProps> {
    state: Todo = {
        id: 0,
        title: '',
        categoryTitle: ''
    };

    handleInputChange = (e: any) => {
        const todoTitle = e.target.value;

        let id = Math.floor(Math.random() * 100);
        this.setState(() => {
            return {
                title: todoTitle,
                id
            };
        });
    };
    handleSelectChange = (e: any) => {
        e.preventDefault();
        const todoCategoryTitle = e.target.value;
        this.setState(() => {
            return {
                categoryTitle: todoCategoryTitle
            };
        });
    };

    onAddTodo = (e: any) => {
        console.log(this.state);
        e.preventDefault();
        this.props.addTodo(this.state);
    };
    render() {
        return (
            <form onSubmit={this.onAddTodo}>
                <label>Todo Title:</label>
                <input type='text' onChange={this.handleInputChange}></input>
                <label>Todo Category:</label>
                <select onChange={this.handleSelectChange}>
                    <option value='null'>Select an option</option>
                    {this.props.categories.map((category, i: number) => {
                        return <option key={i}>{category.name}</option>;
                    })}
                </select>
                <input type='submit' value='Add Todo' />
            </form>
        );
    }
}

export default connect(undefined, { addTodo })(AddTodoForm);
