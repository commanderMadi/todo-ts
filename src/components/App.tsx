import React from 'react';
import { Link } from 'react-router-dom';
class App extends React.Component {
    render() {
        return (
            <div>
                <h1>Welcome to the app!</h1>
                <Link to='/tasks'>Go to the Tasks page</Link>
            </div>
        );
    }
}

export { App };
