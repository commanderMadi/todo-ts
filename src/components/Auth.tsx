import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { Auth } from '../actions';
import { ReduxStoreState } from '../reducers';

interface AuthenticateProps {
  isAuthenticated: Auth;
}

export const requireAuth = (CurrentComponent: React.ComponentType<any>) => {
  class Authenticate extends React.Component<AuthenticateProps> {
    render() {
      const { isAuthenticated } = this.props;
      const { auth } = isAuthenticated;
      return !auth ? (
        <Redirect to='/login' />
      ) : (
        <CurrentComponent {...this.props} />
      );
    }
  }

  const mapStateToProps = (state: ReduxStoreState) => ({
    isAuthenticated: state.isAuthenticated
  });
  return connect(mapStateToProps)(Authenticate);
};

export const publicRoute = (CurrentComponent: React.ComponentType<any>) => {
  class PublicRoute extends React.Component<AuthenticateProps> {
    render() {
      const { isAuthenticated } = this.props;
      const { auth } = isAuthenticated;
      return !auth ? (
        <CurrentComponent {...this.props} />
      ) : (
        <Redirect to='/profile' />
      );
    }
  }

  const mapStateToProps = (state: ReduxStoreState) => ({
    isAuthenticated: state.isAuthenticated
  });
  return connect(mapStateToProps)(PublicRoute);
};
