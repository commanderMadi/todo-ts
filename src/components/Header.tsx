import React from 'react';
import { Link, NavLink } from 'react-router-dom';
import { ReduxStoreState } from '../reducers';
import { Auth, checkAuthentication } from '../actions';
import { connect } from 'react-redux';

export interface HeaderProps {
    isAuthenticated: Auth;
    checkAuthentication: typeof checkAuthentication;
}

export class Header extends React.Component<HeaderProps> {
    onLogoutclick = () => {
        this.props.checkAuthentication(this.props.isAuthenticated);
    };

    render() {
        return (
            <div className='row'>
                <div className='col-sm'>
                    <nav className='navbar navbar-expand-lg navbar-light bg-light'>
                        <Link className='navbar-brand' to='/'>
                            Tasks App
                        </Link>
                        <button
                            className='navbar-toggler'
                            type='button'
                            data-toggle='collapse'
                            data-target='#navbarNav'
                            aria-controls='navbarNav'
                            aria-expanded='false'
                            aria-label='Toggle navigation'
                        >
                            <span className='navbar-toggler-icon'></span>
                        </button>
                        <div
                            className='collapse navbar-collapse'
                            id='navbarNav'
                        >
                            <ul id='nav-mobile' className='navbar-nav'>
                                <li className='nav-item'>
                                    <NavLink className='nav-link' to='/'>
                                        Home
                                    </NavLink>
                                </li>
                                <li className='nav-item'>
                                    <NavLink className='nav-link' to='/tasks'>
                                        Tasks
                                    </NavLink>
                                </li>
                                <li className='nav-item'>
                                    <NavLink className='nav-link' to='/profile'>
                                        Profile
                                    </NavLink>
                                </li>

                                {!this.props.isAuthenticated.auth && (
                                    <li className='nav-item'>
                                        <NavLink
                                            className='nav-link text-success'
                                            to='/login'
                                        >
                                            Login
                                        </NavLink>
                                    </li>
                                )}
                            </ul>
                            {this.props.isAuthenticated.auth && (
                                <button
                                    className='btn btn-sm btn-danger ml-auto'
                                    onClick={this.onLogoutclick}
                                >
                                    Logout
                                </button>
                            )}
                        </div>
                    </nav>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state: ReduxStoreState) => {
    return {
        isAuthenticated: state.isAuthenticated
    };
};

export default connect(mapStateToProps, { checkAuthentication })(Header);
