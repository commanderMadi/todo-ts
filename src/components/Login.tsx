import React from 'react';
import { connect } from 'react-redux';
import { ReduxStoreState } from '../reducers';
import { checkAuthentication, Auth } from '../actions';

export interface LoginProps {
    isAuthenticated: Auth;
    checkAuthentication: typeof checkAuthentication;
}

interface LoginState {
    username: string;
    password: string;
    error: null | string;
}

export class Login extends React.Component<LoginProps, LoginState> {
    constructor(props: LoginProps) {
        super(props);
        this.state = { username: 'Admin', password: '54321', error: null };
        this.onFormSubmit = this.onFormSubmit.bind(this);
    }

    componentDidMount() {
        const loginData = JSON.stringify(this.state);
        localStorage.setItem('credentials', loginData);
    }

    checkCredentials(username: string, password: string) {
        const data: any = localStorage.getItem('credentials');
        const parsedData = JSON.parse(data);

        if (
            username === parsedData.username &&
            password === parsedData.password
        ) {
            this.props.checkAuthentication(this.props.isAuthenticated);
        } else {
            this.setState(() => {
                return {
                    error: 'Wrong username or password.'
                };
            });
        }
    }

    onFormSubmit(e: any) {
        e.preventDefault();
        const [username, password] = e.target.elements;
        this.checkCredentials(username.value, password.value);
    }

    render() {
        return (
            <div>
                {this.state.error && (
                    <p className='col-md-12 text-danger mt-4'>
                        {this.state.error}
                    </p>
                )}
                <form onSubmit={this.onFormSubmit}>
                    <div className='form-group'>
                        <label htmlFor='user'>Username</label>
                        <input
                            type='text'
                            className='form-control'
                            id='usernameinput'
                        />
                    </div>
                    <div className='form-group'>
                        <label htmlFor='pass'>Password</label>
                        <input
                            type='password'
                            className='form-control'
                            id='passwordinput'
                        />
                    </div>
                    <div className='form-group'>
                        <input
                            value='Log In'
                            type='submit'
                            className='btn btn-primary'
                        />
                    </div>
                </form>
            </div>
        );
    }
}

const mapStateToProps = (state: ReduxStoreState) => {
    return {
        isAuthenticated: state.isAuthenticated
    };
};

export default connect(mapStateToProps, { checkAuthentication })(Login);
