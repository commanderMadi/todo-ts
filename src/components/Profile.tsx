import React from 'react';
import { Link } from 'react-router-dom';

class Profile extends React.Component {
  render() {
    return (
      <div>
        <p>Welcome, Admin</p>
        <Link to='/tasks'>Go to the Tasks page</Link>
      </div>
    );
  }
}

export { Profile };
