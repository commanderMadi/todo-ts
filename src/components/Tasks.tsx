import React from 'react';
import { ReduxStoreState } from '../reducers/';
import { connect } from 'react-redux';
import { Task, Category, getCategories, deleteTask, addTask } from '../actions';
import { AddTaskForm } from '../components/AddTaskForm';

export interface TasksProps {
    categories: Category[];
    getCategories: Function;
    deleteTask: typeof deleteTask;
    addTask: typeof addTask;
}

export class Tasks extends React.Component<TasksProps> {
    constructor(props: TasksProps) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
    }
    componentDidMount() {
        this.props.getCategories();
    }

    onDelete(id: string) {
        this.props.deleteTask(id);
    }

    onSubmit(newTask: Task) {
        this.props.addTask(newTask);
    }

    renderCategories() {
        return (
            this.props.categories &&
            this.props.categories.map((category: Category, index: number) => {
                return (
                    <div className='col-md-6' key={index}>
                        <h3>{category.name}</h3>

                        {category.tasks.map((task: Task) => {
                            return (
                                <div className='mb-3' key={task.id}>
                                    <span className='pr-4'>{task.title}</span>
                                    <button
                                        className='task_remove btn btn-sm btn-danger'
                                        onClick={() => this.onDelete(task.id)}
                                    >
                                        Remove Task
                                    </button>
                                </div>
                            );
                        })}
                    </div>
                );
            })
        );
    }
    render() {
        return (
            <div>
                <AddTaskForm
                    onSubmit={this.onSubmit}
                    categories={this.props.categories}
                />
                <div className='row mt-5'>
                    <div className='col-12 mb-4'>
                        <h2>Tasks Categories</h2>
                    </div>
                    {this.props.categories.length > 0 ? (
                        this.renderCategories()
                    ) : (
                        <p className='no_data'>No data</p>
                    )}
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state: ReduxStoreState) => {
    return {
        categories: state.categories
    };
};

export default connect(mapStateToProps, { getCategories, addTask, deleteTask })(
    Tasks
);
