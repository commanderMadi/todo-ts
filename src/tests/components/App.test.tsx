import React from 'react';
import { App } from '../../components/App';
import { shallow } from 'enzyme';

test('App component should load correctly', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
});
