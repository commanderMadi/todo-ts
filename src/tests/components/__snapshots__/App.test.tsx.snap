// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should load correctly 1`] = `
<div>
  <h1>
    Welcome to the app!
  </h1>
</div>
`;
